File: apps/desktop/src/main/indexing.ts
/**
 * Google Indexing API 클라이언트 — 외부 라이브러리 없이 Node crypto 만 사용.
 *
 * 흐름: 서비스계정 JSON → RS256 JWT 서명 → oauth2 토큰 교환 → urlNotifications:publish.
 * 서비스계정 키가 없으면 isConfigured()=false 로 안전하게 비활성(크래시 X).
 *
 * 주의: Indexing API 는 공식적으로 JobPosting/BroadcastEvent 페이지용이며,
 * 일반 페이지 제출은 구글 약관상 권장 용도가 아니다(회색지대). 서비스계정 이메일을
 * 각 사이트 Search Console 속성에 "소유자"로 추가해야 실제 제출이 동작한다.
 */

import { createSign } from "crypto";

const TOKEN_URL = "https://oauth2.googleapis.com/token";
const PUBLISH_URL = "https://indexing.googleapis.com/v3/urlNotifications:publish";
const SCOPE = "https://www.googleapis.com/auth/indexing";

export interface ServiceAccount {
  client_email: string;
  private_key: string;
  token_uri?: string;
}

export type NotifyType = "URL_UPDATED" | "URL_DELETED";

function base64url(input: Buffer | string): string {
  const buf = typeof input === "string" ? Buffer.from(input, "utf8") : input;
  return buf.toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

/** 서비스계정 JSON 문자열을 파싱·검증. 실패 시 명확한 에러. */
export function parseServiceAccount(jsonText: string | null | undefined): ServiceAccount {
  if (!jsonText || !jsonText.trim()) {
    throw new Error("서비스계정 키가 설정되지 않았습니다(미설정).");
  }
  let obj: Record<string, unknown>;
  try {
    obj = JSON.parse(jsonText);
  } catch {
    throw new Error("서비스계정 JSON 형식이 올바르지 않습니다.");
  }
  const email = typeof obj.client_email === "string" ? obj.client_email : "";
  const key = typeof obj.private_key === "string" ? obj.private_key : "";
  if (!email || !key) {
    throw new Error("서비스계정 JSON 에 client_email / private_key 가 없습니다.");
  }
  return { client_email: email, private_key: key, token_uri: typeof obj.token_uri === "string" ? obj.token_uri : TOKEN_URL };
}

/** 키가 유효하게 설정돼 있으면 true(네트워크 호출 없음). */
export function isConfigured(jsonText: string | null | undefined): boolean {
  try {
    parseServiceAccount(jsonText);
    return true;
  } catch {
    return false;
  }
}

/** 서비스계정으로 서명한 JWT(assertion). nowSec 주입 가능(테스트용). */
export function buildAssertion(sa: ServiceAccount, nowSec: number = Math.floor(Date.now() / 1000)): string {
  const header = base64url(JSON.stringify({ alg: "RS256", typ: "JWT" }));
  const claim = base64url(JSON.stringify({
    iss: sa.client_email,
    scope: SCOPE,
    aud: sa.token_uri || TOKEN_URL,
    iat: nowSec,
    exp: nowSec + 3600,
  }));
  const signingInput = `${header}.${claim}`;
  const signature = createSign("RSA-SHA256").update(signingInput).end().sign(sa.private_key);
  return `${signingInput}.${base64url(signature)}`;
}

/** JWT → access_token 교환. */
export async function getAccessToken(sa: ServiceAccount): Promise<string> {
  const assertion = buildAssertion(sa);
  const body = new URLSearchParams({
    grant_type: "urn:ietf:params:oauth:grant-type:jwt-bearer",
    assertion,
  });
  const res = await fetch(sa.token_uri || TOKEN_URL, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: body.toString(),
  });
  const json = (await res.json().catch(() => ({}))) as Record<string, unknown>;
  if (!res.ok || typeof json.access_token !== "string") {
    const desc = (json.error_description || json.error || `HTTP ${res.status}`) as string;
    throw new Error(`토큰 발급 실패: ${desc}`);
  }
  return json.access_token;
}

export interface SubmitResult {
  url: string;
  ok: boolean;
  error?: string;
}

/** 단일 URL 색인 알림 제출. */
export async function submitUrl(accessToken: string, url: string, type: NotifyType = "URL_UPDATED"): Promise<SubmitResult> {
  try {
    const res = await fetch(PUBLISH_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json", Authorization: `Bearer ${accessToken}` },
      body: JSON.stringify({ url, type }),
    });
    if (!res.ok) {
      const json = (await res.json().catch(() => ({}))) as Record<string, unknown>;
      const err = json.error as { message?: string } | undefined;
      return { url, ok: false, error: err?.message || `HTTP ${res.status}` };
    }
    return { url, ok: true };
  } catch (e) {
    return { url, ok: false, error: (e as Error).message };
  }
}

/** 템플릿({domain}/{slug})으로 발행 URL 생성. */
export function buildPostUrl(template: string, domain: string, slug: string): string {
  const t = template?.trim() || "https://{domain}/{slug}";
  return t.replace(/\{domain\}/g, domain).replace(/\{slug\}/g, slug);
}

File: runtime/indexing.py
"""Google Indexing API 클라이언트 (TS indexing.ts 포팅).

서비스계정 JSON → RS256 JWT(cryptography) → oauth2 토큰 → urlNotifications:publish.
HTTP는 stdlib urllib 사용. 키 없으면 is_configured()=False 로 안전 비활성.

주의: Indexing API 는 공식적으로 JobPosting/BroadcastEvent 용이며 일반 페이지 제출은
구글 약관상 회색지대. 서비스계정 이메일을 각 사이트 Search Console 에 소유자로 추가해야 동작.
"""

from __future__ import annotations

import base64
import json
import time
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding

TOKEN_URL = "https://oauth2.googleapis.com/token"
PUBLISH_URL = "https://indexing.googleapis.com/v3/urlNotifications:publish"
SCOPE = "https://www.googleapis.com/auth/indexing"


@dataclass
class ServiceAccount:
    client_email: str
    private_key: str
    token_uri: str = TOKEN_URL


def _b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def parse_service_account(json_text: str | None) -> ServiceAccount:
    if not json_text or not json_text.strip():
        raise ValueError("서비스계정 키가 설정되지 않았습니다(미설정).")
    try:
        obj = json.loads(json_text)
    except json.JSONDecodeError:
        raise ValueError("서비스계정 JSON 형식이 올바르지 않습니다.")
    email = obj.get("client_email")
    key = obj.get("private_key")
    if not isinstance(email, str) or not isinstance(key, str) or not email or not key:
        raise ValueError("서비스계정 JSON 에 client_email / private_key 가 없습니다.")
    return ServiceAccount(client_email=email, private_key=key,
                          token_uri=obj.get("token_uri") or TOKEN_URL)


def is_configured(json_text: str | None) -> bool:
    try:
        parse_service_account(json_text)
        return True
    except Exception:
        return False


def build_assertion(sa: ServiceAccount, now_sec: int | None = None) -> str:
    now = now_sec if now_sec is not None else int(time.time())
    header = _b64url(json.dumps({"alg": "RS256", "typ": "JWT"}, separators=(",", ":")).encode())
    claim = _b64url(json.dumps({
        "iss": sa.client_email, "scope": SCOPE, "aud": sa.token_uri or TOKEN_URL,
        "iat": now, "exp": now + 3600,
    }, separators=(",", ":")).encode())
    signing_input = f"{header}.{claim}".encode("ascii")
    private_key = serialization.load_pem_private_key(sa.private_key.encode("utf-8"), password=None)
    signature = private_key.sign(signing_input, padding.PKCS1v15(), hashes.SHA256())
    return f"{header}.{claim}.{_b64url(signature)}"


def get_access_token(sa: ServiceAccount, *, timeout: int = 20) -> str:
    body = urllib.parse.urlencode({
        "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
        "assertion": build_assertion(sa),
    }).encode("ascii")
    req = urllib.request.Request(
        sa.token_uri or TOKEN_URL, data=body,
        headers={"Content-Type": "application/x-www-form-urlencoded"}, method="POST")
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            data = json.loads(resp.read().decode())
    except urllib.error.HTTPError as e:
        detail = e.read().decode(errors="ignore")
        raise RuntimeError(f"토큰 발급 실패: HTTP {e.code} {detail[:200]}")
    token = data.get("access_token")
    if not token:
        raise RuntimeError(f"토큰 발급 실패: {data}")
    return token


def submit_url(access_token: str, url: str, notify_type: str = "URL_UPDATED",
               *, timeout: int = 20) -> dict:
    payload = json.dumps({"url": url, "type": notify_type}).encode()
    req = urllib.request.Request(
        PUBLISH_URL, data=payload,
        headers={"Content-Type": "application/json", "Authorization": f"Bearer {access_token}"},
        method="POST")
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            resp.read()
        return {"url": url, "ok": True}
    except urllib.error.HTTPError as e:
        detail = e.read().decode(errors="ignore")
        try:
            msg = json.loads(detail).get("error", {}).get("message", f"HTTP {e.code}")
        except Exception:
            msg = f"HTTP {e.code}"
        return {"url": url, "ok": False, "error": msg}
    except Exception as e:  # noqa: BLE001
        return {"url": url, "ok": False, "error": str(e)}


def build_post_url(template: str, domain: str, slug: str) -> str:
    t = (template or "").strip() or "https://{domain}/{slug}"
    return t.replace("{domain}", domain).replace("{slug}", slug)
